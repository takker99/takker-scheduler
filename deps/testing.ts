export * from "https://deno.land/std@0.198.0/assert/mod.ts";
