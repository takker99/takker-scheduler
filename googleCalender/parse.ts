import type { Event } from "./events.ts";
import { parse as parseTask } from "../task.ts";
import { addSeconds } from "../deps/date-fns.ts";
import { encodeTitle } from "../lib/utils.ts";

interface ParseInit {
  /** 取得先project */ project: string;
  /** 取得先page title */ title: string;
  /** 行ID */ id: string;
}
/** 指定したテキストから予定と記録のcalendar登録用event objectsを作る
 *
 * @param text - task line (改行不可)
 * @param init - 追加のproperties
 */
export function parse(
  text: string,
  init: ParseInit,
): [Event | undefined, Event | undefined] | undefined {
  const task = parseTask(text);
  if (!task) return;

  const { title, plan, record } = task;
  const source = {
    title: init.title,
    url: `https://scrapbox.io/${init.project}/${
      encodeTitle(init.title)
    }#${init.id}`,
  };

  // eventの名前から[]を外して末尾の日付を消しておく
  const summary = title
    .replace(/\[([^\]]+)\]/g, "$1")
    .trim()
    .replace(/\s\d{4}-\d{2}-\d{2}$/, "")
    // 前に日付があるタイプも消す
    .replace(/^\d{4}-\d{2}-\d{2}\s/, "");

  return [
    plan.start && plan.duration && summary !== ""
      ? {
        summary,
        start: plan.start,
        end: addSeconds(plan.start, plan.duration),
        source,
        id: init.id,
      }
      : // 名無しの予定は飛ばす
        undefined,
    record.start && record.end
      ? {
        summary,
        start: record.start,
        end: record.end,
        source,
        id: init.id,
      }
      : undefined,
  ];
}
