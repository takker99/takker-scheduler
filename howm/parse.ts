import { isValid } from "../deps/date-fns.ts";
import { Result } from "../deps/scrapbox-std.ts";
import {
  format,
  fromDate,
  isBefore,
  LocalDate,
  LocalDateTime,
  toDate,
} from "./localDate.ts";
import { fromStatus, Status, toStatus } from "./status.ts";

export const toFrequency = (
  symbol: string,
): "yearly" | "monthly" | "weekly" | "daily" | undefined => {
  switch (symbol.toLowerCase()) {
    case "y":
      return "yearly";
    case "m":
      return "monthly";
    case "w":
      return "weekly";
    case "d":
      return "daily";
    default:
      return;
  }
};

export interface Repeat {
  type: "yearly" | "monthly" | "weekly" | "daily";
  count: number;
}

/* いずれ実装する
/** 曜日指定情報を加えたDate *
export interface ExDate {
  base: Date;
  startOfDay: 0 | 1 | 2 | 3 | 4 | 5 | 6;
  weekNum: number;
}
*/

export interface Task {
  /** task name */
  name: string;

  /** howm記号 */
  status: Status;

  /** hown記号のオプション */
  speed?: number;

  /** 開始日時 */
  start: LocalDate | LocalDateTime;

  /** 終了日時 */
  end?: LocalDate | LocalDateTime;

  /** 所要時間 (単位はmin) **/
  duration?: number;

  /** 解析前の文字列 */
  raw: string;
}

export interface TaskRangeError {
  name: "TaskRangeError";
  message: string;
}

export interface InvalidDateError {
  name: "InvalidDateError";
  message: string;
}

/** Taskを解析する
 *
 * @param text Taskの文字列
 * @return 解析結果。Taskでなければ`undefined`を返す
 */
export const parse = (
  text: string,
): Result<Task, TaskRangeError | InvalidDateError> | undefined => {
  const matched = text.match(
    /([\+\-!~.])(\d+)?@(\d{4})-(\d{2})-(\d{2})(?:T(\d{2}):(\d{2}))?(?:\/(\d{2}):(\d{2})|\/(?:(?:(?:(\d{4})-)?(\d{2})-)?(\d{2})(?:T(\d{2}):(\d{2}))?)|D(\d+))?/i,
  );
  if (!matched) return;

  const [
    matchedText,
    statusSym,
    speedStr,
    syear,
    smonth,
    sdate,
    shours,
    sminutes,
    ehours2,
    eminutes2,
    eyear,
    emonth,
    edate,
    ehours,
    eminutes,
    durationStr,
  ] = matched;

  /** task name */
  const name = `${text.slice(0, matched.index).trim()}${
    text.slice((matched.index ?? 0) + matchedText.length).trim()
  }`;

  const status_ = toStatus(statusSym);
  // /[\+\-!~.]/以外にマッチするはずがない
  if (!status_) throw Error("status must be +,,-,!,~,.");

  // 後方互換性用コード
  // 先頭が✅か❌のときは、`done`とみなす
  /** task status */
  const status = /^[\uFE00-\uFE0F]*[✅❌]/.test(text) ? "done" : status_;
  const start_: LocalDate = {
    year: parseInt(syear),
    month: parseInt(smonth),
    date: parseInt(sdate),
  };
  /** 開始日時 */
  const start: LocalDate | LocalDateTime = !shours
    ? start_
    : { ...start_, hours: parseInt(shours), minutes: parseInt(sminutes) };
  if (!isValid(toDate(start))) {
    return {
      ok: false,
      value: {
        name: "InvalidDateError",
        message: `The start of the task "${format(start)}" is an invalid date.`,
      },
    };
  }

  const task: Task = { name, status, start, raw: text };
  if (speedStr) task.speed = parseInt(speedStr);
  if (ehours2 || edate) {
    let end: LocalDate | LocalDateTime = {
      year: start.year,
      month: start.month,
      date: start.date,
    };
    if (ehours2) {
      end = { ...end, hours: parseInt(ehours2), minutes: parseInt(eminutes2) };
    } else {
      if (eyear) end.year = parseInt(eyear);
      if (emonth) end.month = parseInt(emonth);
      if (edate) end.date = parseInt(edate);
      if (ehours) {
        end = {
          ...end,
          hours: parseInt(ehours),
          minutes: parseInt(eminutes),
        };
      }
    }
    if (!isValid(toDate(end))) {
      return {
        ok: false,
        value: {
          name: "InvalidDateError",
          message: `The end of the task "${format(end)}" is an invalid date.`,
        },
      };
    }
    if (isBefore(end, start)) {
      return {
        ok: false,
        value: {
          name: "TaskRangeError",
          message: `The start of an task cannot be after its end.\n\nstart:${
            format(start)
          }\nend:${format(end)}`,
        },
      };
    }
    task.end = end;
  }
  if (durationStr) task.duration = parseInt(durationStr);

  return { ok: true, value: task };
};

/** 終日タスクかどうか判定する
 *
 * `start`に時刻が含まれていないタスクは全て終日タスクだとみなす
 */
export const isAllDay = (task: Pick<Task, "start">): boolean =>
  !("hours" in task.start);

/** タスクの所要時間を分単位で得る
 *
 * 所要時間が設定されていない場合は`undefined`を返す
 */
export const getDuration = (
  task: Pick<Task, "start" | "end" | "duration">,
): number | undefined =>
  task.end
    ? isAllDay(task) ? undefined : Math.round(
      (toDate(task.end).getTime() - toDate(task.start).getTime()) / (60 * 1000),
    )
    : task.duration;

/** タスクの終了日時を得る
 *
 * 終日の場合は、最終日の翌日0時を終了日時とする
 */
export const getEnd = (task: Task): LocalDateTime => {
  if (task.end) {
    if ("hours" in task.end) return task.end;
    const end = toDate(task.end);
    end.setDate(end.getDate() + 1);
    return fromDate(end);
  }
  if (task.duration === undefined || !("hours" in task.start)) {
    const end = toDate(task.start);
    end.setHours(0);
    end.setMinutes(0);
    end.setDate(end.getDate() + 1);
    return fromDate(end);
  }
  const end = toDate(task.start);
  end.setMinutes(end.getMinutes() + task.duration);
  return fromDate(end);
};

/** Task objectからタスクリンクの文字列を作る */
export const toString = (task: Task): string => {
  const duration = getDuration(task);

  return `${fromStatus(task.status)}${task.speed ?? ""}@${format(task.start)}${
    duration === undefined ? "" : `D${duration}`
  }${task.name}`;
};
