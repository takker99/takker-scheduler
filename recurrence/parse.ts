import { addMinutes, isValid } from "../deps/date-fns.ts";
import { Result, isNumber } from "../deps/scrapbox-std.ts";
import { format, fromDate, LocalDateTime, toDate } from "../howm/localDate.ts";
import {
    Task,

} from "../howm/parse.ts";
import {
    Event,
} from "../event/parse.ts";
import { Status, toStatus } from "../howm/status.ts";

/** 繰り返しを表す */
export interface Recurrence {
    /** task name */
    name: string;

    /** 開始年 */
    year: number | "*";

    /** 開始月 */
    month: number | "*";

    /** 開始日 */
    date: RegDate

    hours: number;

    minutes: number;

    /** 所要時間 */
    duration: number;

    status?: Status;

    /** 解析前の文字列 */
    raw: string;
}

export type RegDate = number | "*" | "Sun" | "Mon" | "Tue" | "Wed" | "Thu" | "Fri" | "Sat";

export interface LackDateError {
    name: "LackDateError";
    message: string;
}

/** Eventを解析する
 *
 * @param text Eventの文字列
 * @return 解析結果。Eventでなければ`undefined`を返す
 */
export const parse = (
    text: string,
):
    Recurrence | undefined => {
    const matched = text.match(
        /([\+\-!~]?)@(\d{4}|\*)-(\d{2}|\*)-(\d{2}|\*|Sun|Mon|Tue|Wed|Thu|Fri|Sat)T(\d{2}):(\d{2})D(\d+)/i
    );
    // eventでないとき
    if (!matched) return;

    const [
        matchedText,
        statusStr,
        syear,
        smonth,
        sdate,
        shours,
        sminutes,
        durationStr,
    ] = matched;

    /** task name */
    const name = `${text.slice(0, matched.index).trim()}${text.slice((matched.index ?? 0) + matchedText.length).trim()
        }`;
    const status = toStatus(statusStr);
    const year = checkRegNumber(syear);
    const month = checkRegNumber(smonth)
    const date = checkRegDate(sdate)
    const hours = parseInt(shours);
    const minutes = parseInt(sminutes)

    const recurrence: Recurrence = { name, year, month, date, hours, minutes, duration: parseInt(durationStr), raw: text };
    if (status) recurrence.status = status



    return recurrence
};

/** 指定日に繰り返す繰り返しタスクか調べる。
 *
 * 繰り返す場合はそのタスクを指定日の日付で生成する。
 * 繰り返さない場合は`undefined`を返す
 */
export const makeRepeat = (recurrence: Recurrence, date: Date): Task | Event | undefined => {
    const localDate = fromDate(date);
    if (isNumber(recurrence.year) && recurrence.year !== date.getFullYear()) return;
    if (isNumber(recurrence.month) && recurrence.month !== date.getMonth() + 1) return;
    if (isNumber(recurrence.date) && recurrence.date !== date.getDate()) return;
    if (!isNumber(recurrence.date) && recurrence.date !== "*" && toDayNumber(recurrence.date) !== date.getDay()) return;

    return recurrence.status ? {
        name: recurrence.name,
        status: recurrence.status,
        start: localDate,
        duration: recurrence.duration,
        raw: recurrence.raw,
    } : {
        name: recurrence.name,
        start: localDate,
        end: fromDate(addMinutes(date, recurrence.duration)),
        raw: recurrence.raw,
    }
}

/** 数字か`*`のときのみ使える */
const checkRegNumber = (text: string): number | "*" => {
    if (text === "*") return text;
    return parseInt(text);
}

/** エラー処理なし */
const checkRegDate = (text: string): RegDate => {
    switch (text.toLowerCase()) {
        case "*": return "*";
        case "sun": return "Sun"
        case "mon": return "Mon"
        case "tue": return "Tue"
        case "wed": return "Wed"
        case "thu": return "Thu"
        case "fri": return "Fri"
        case "sat": return "Sat"
        default: return parseInt(text)
    }
}

const toDayNumber = (day:
    "Sun" | "Mon" | "Tue" | "Wed" | "Thu" | "Fri" | "Sat"

): 0 | 1 | 2 | 3 | 4 | 5 | 6 => {

    switch (day) {
        case "Sun": return 0
        case "Mon": return 1
        case "Tue": return 2
        case "Wed": return 3
        case "Thu": return 4
        case "Fri": return 5
        case "Sat": return 6
    }
}