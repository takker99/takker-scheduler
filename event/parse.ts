import {
  addMinutes,
  differenceInCalendarDays,
  differenceInCalendarMonths,
  isValid,
} from "../deps/date-fns.ts";
import { Result } from "../deps/scrapbox-std.ts";
import { format, fromDate, LocalDateTime, toDate } from "../howm/localDate.ts";
import {
  getDuration as getDurationOfTask,
  InvalidDateError,
  parse as parseTask,
  Task,
  TaskRangeError,
} from "../howm/parse.ts";
import { Recurrence, toFrequency } from "./recurrence.ts";

/** 予定 (開始日時と終了日時が確定したアイテム)を表す */
export interface Event {
  /** task name */
  name: string;

  /** 開始日時 */
  start: LocalDateTime;

  /** 終了日時 */
  end: LocalDateTime;

  /** 解析した文字列がタスクでもあった場合は、そのデータを入れる */
  task?: Task;

  recurrence?: Recurrence;

  /** 解析前の文字列 */
  raw: string;
}

export interface LackDateError {
  name: "LackDateError";
  message: string;
}

/** Eventを解析する
 *
 * @param text Eventの文字列
 * @return 解析結果。Eventでなければ`undefined`を返す
 */
export const parse = (
  text: string,
):
  | Result<Event | Task, TaskRangeError | InvalidDateError | LackDateError>
  | undefined => {
  // 省略されたデータはタスクから流用する
  const task = parseTask(text);
  if (task?.ok === false) return task;

  const taskName = task?.value?.name ?? text;
  const matched = taskName.match(
    /@(?:(\d{2}):(\d{2})|(?:(?:(\d{4})-)?(\d{2})-)?(\d{2})(?:T(\d{2}):(\d{2}))?)?(?:\/(\d{2}):(\d{2})|\/(?:(?:(?:(\d{4})-)?(\d{2})-)?(\d{2})(?:T(\d{2}):(\d{2}))?)|D(\d+))?(?:R([YMWD])?(\d+))?/i,
  );
  // eventでないとき
  if (!matched) return task;

  const [
    matchedText,
    shours2,
    sminutes2,
    syear,
    smonth,
    sdate,
    shours,
    sminutes,
    ehours2,
    eminutes2,
    eyear,
    emonth,
    edate,
    ehours,
    eminutes,
    durationStr,
    symbol,
    countStr,
  ] = matched;

  /** task name */
  const name = `${taskName.slice(0, matched.index).trim()}${
    taskName.slice((matched.index ?? 0) + matchedText.length).trim()
  }`;

  const sresult = checkStart(
    shours2,
    sminutes2,
    syear,
    smonth,
    sdate,
    shours,
    sminutes,
    task?.value,
  );
  if (!sresult.ok) return sresult;
  /** 開始日時 */
  const start = sresult.value;
  if (!isValid(toDate(start))) {
    return {
      ok: false,
      value: {
        name: "InvalidDateError",
        message: `The start of the task "${format(start)}" is an invalid date.`,
      },
    };
  }

  const eresult = checkEnd(
    ehours2,
    eminutes2,
    eyear,
    emonth,
    edate,
    ehours,
    eminutes,
    durationStr,
    start,
    task?.value,
  );
  if (!eresult.ok) return eresult;
  /** 終了日時 */
  const end = eresult.value;
  if (!isValid(toDate(end))) {
    return {
      ok: false,
      value: {
        name: "InvalidDateError",
        message: `The end of the task "${format(end)}" is an invalid date.`,
      },
    };
  }

  const event: Event = { name, start, end, raw: text };
  if (task?.value) event.task = task.value;
  if (countStr) {
    const count = parseInt(countStr);
    const frequency = toFrequency(symbol ?? "D") ?? "daily";
    event.recurrence = { frequency, count };
  }

  return { ok: true, value: event };
};

const checkStart = (
  shours2: string,
  sminutes2: string,
  syear: string,
  smonth: string,
  sdate: string,
  shours: string,
  sminutes: string,
  task?: Task,
): Result<LocalDateTime, LackDateError> => {
  if (shours2) {
    if (!task) {
      return {
        ok: false,
        value: {
          name: "LackDateError",
          message:
            "No start date found in both a task specifier and an event specifier.",
        },
      };
    }
    return {
      ok: true,
      value: {
        year: task.start.year,
        month: task.start.month,
        date: task.start.date,
        hours: parseInt(shours2),
        minutes: parseInt(sminutes2),
      },
    };
  }

  const year = syear ? parseInt(syear) : task?.start?.year;
  const month = smonth ? parseInt(smonth) : task?.start?.month;
  const date = sdate ? parseInt(sdate) : task?.start?.date;
  const hours = shours
    ? parseInt(shours)
    : task?.start && ("hours" in task.start)
    ? task.start.hours
    : undefined;
  const minutes = sminutes
    ? parseInt(sminutes)
    : task?.start && ("minutes" in task.start)
    ? task.start.minutes
    : undefined;
  const target = year === undefined
    ? "year"
    : month === undefined
    ? "month"
    : date === undefined
    ? "date"
    : hours === undefined
    ? "hours"
    : minutes === undefined
    ? "minutes"
    : false;
  if (target) {
    return {
      ok: false,
      value: {
        name: "LackDateError",
        message: `Could not found any "${target}" parameter.`,
      },
    };
  }

  return {
    ok: true,
    value: {
      year: year!,
      month: month!,
      date: date!,
      hours: hours!,
      minutes: minutes!,
    },
  };
};

const checkEnd = (
  ehours2: string,
  eminutes2: string,
  eyear: string,
  emonth: string,
  edate: string,
  ehours: string,
  eminutes: string,
  durationStr: string,
  start: LocalDateTime,
  task?: Task,
): Result<LocalDateTime, LackDateError> => {
  // 終了時刻が設定されているとき
  if (ehours2) {
    return {
      ok: true,
      value: {
        year: start.year,
        month: start.month,
        date: start.date,
        hours: parseInt(ehours2),
        minutes: parseInt(eminutes2),
      },
    };
  }
  // 所要時間が設定されているとき
  if (durationStr) {
    return {
      ok: true,
      value: fromDate(addMinutes(toDate(start), parseInt(durationStr))),
    };
  }
  // eventから終了日時を計算できない場合
  if (!edate) {
    // Taskから所要時間を取得できればそれを使う
    if (task) {
      const duration = getDurationOfTask(task);
      if (duration !== undefined) {
        return {
          ok: true,
          value: fromDate(addMinutes(toDate(start), duration)),
        };
      }
    }
    // 取得できなければエラーを返す
    return {
      ok: false,
      value: {
        name: "LackDateError",
        message:
          "No end date found in both a task specifier and an event specifier.",
      },
    };
  }

  const year = eyear ? parseInt(eyear) : start.year;
  const month = emonth ? parseInt(emonth) : start.month;
  const date = edate ? parseInt(edate) : start.date;
  const hours = ehours ? parseInt(ehours) : start.hours;
  const minutes = eminutes ? parseInt(eminutes) : start.minutes;

  return { ok: true, value: { year, month, date, hours, minutes } };
};

/** 指定日に繰り返す繰り返しタスクか調べる。
 *
 * 繰り返す場合はそのタスクを指定日の日付で生成する。
 * 繰り返さない場合は`undefined`を返す
 */
export const makeRepeat = (
  event: Event,
  date: Date,
): Task | Event | undefined => {
  if (!event.recurrence) return;
  const localDate = fromDate(date);

  const recurrence = event.recurrence;
  // 繰り返す場合のみ通過させる
  switch (event.recurrence.frequency) {
    case "yearly": {
      // 間隔があっているかチェック
      if (
        Math.abs(localDate.year - event.start.year) %
            (recurrence.count ?? 1) !== 0
      ) return;

      // 日と月が一致しているかチェック
      if (
        event.start.month !== localDate.month ||
        event.start.date !== localDate.date
      ) return;

      break;
    }
    case "monthly": {
      /** 与えられた日付とeventの繰り返し起点との月差 */
      const diff = differenceInCalendarMonths(
        toDate(localDate),
        toDate(event.start),
      );
      if (diff % (recurrence.count ?? 1) !== 0) return;

      break;
    }
    case "weekly":
    case "daily": {
      const interval = recurrence.frequency === "weekly" ? 7 : 1;
      const diff = differenceInCalendarDays(
        toDate(localDate),
        toDate(event.start),
      );
      if (diff % ((recurrence.count ?? 1) * interval) !== 0) return;

      break;
    }
  }

  // dateの日付を使ってTask or Eventを生成する
  const newEvent = structuredClone(event);

  newEvent.start.year = localDate.year;
  const duration = getDuration(newEvent);
  if (newEvent.task) {
    const task = newEvent.task;
    delete task.end;
    task.duration = duration;
    return task;
  }
  newEvent.end = fromDate(addMinutes(toDate(newEvent.start), duration));
  return newEvent;
};

/** Eventの長さを分単位で返す */
export const getDuration = (event: Event): number =>
  Math.round(
    (toDate(event.end).getTime() - toDate(event.start).getTime()) / 60 * 1000,
  );
