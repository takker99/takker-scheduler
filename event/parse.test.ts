/// <reference lib="deno.ns" />

import { parse } from "./parse.ts";
import { assertSnapshot } from "../deps/testing.ts";

const runTests = async (t: Deno.TestContext, args: string[]) => {
  for (const arg of args) {
    await t.step(arg, () => assertSnapshot(t, parse(arg)));
  }
};

Deno.test("parse()", async (t) => {
  await t.step("howm taskのみのときは`Task`で返す", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな~@2023-09-12",
    ]));
  await t.step("taskがないときは終了時刻まで明記する", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな@2023-09-12T12:50D10",
      "ハイウェイ惑星 買おうかな@2023-09-12T12:50/14:20",
      "終了時刻がないのでエラーになる@2023-09-12T12:50",
      "開始時刻がないのでエラー@2023-09-12D10",
    ]));
  await t.step("taskの日付情報から継承する", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13",
      "ハイウェイ惑星 買おうかな~@2023-09-12D10@13T12:50",
      "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50D10",
      "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50/14:20",
      "ハイウェイ惑星 買おうかな~@2023-09-12@12-13T12:50D10",
      "ハイウェイ惑星 買おうかな~@2023-09-12@2024-12-13T12:50D10",
      "時刻がない~@2023-09-12@13",
      "終了時刻がない~@2023-09-12@13T13:45",
      "開始時刻がない~@2023-09-12@13D45",
    ]));
  await t.step("taskとeventの日付指定は順不同", (t) =>
    runTests(t, [
      "@13ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10",
      "~@2023-09-12T12:50D10 ←がTaskを指定し、~@09-24←がEventを指定する",
    ]));
  await t.step("繰り返し", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13R4",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13R47",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13R",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13RM8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13RY8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13RW8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13RD8",
      "読書会@2023-09-12T12:50D10@13RD8",
    ]));
});
