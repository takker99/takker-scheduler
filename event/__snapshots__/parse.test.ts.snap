export const snapshot = {};

snapshot[`parse() > howm taskのみのときは\`Task\`で返す 1`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12",
    start: {
      date: 12,
      month: 9,
      year: 2023,
    },
    status: "up-down",
  },
}
`;

snapshot[`parse() > taskがないときは終了時刻まで明記する 1`] = `
{
  ok: true,
  value: {
    end: {
      date: 12,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな@2023-09-12T12:50D10",
    start: {
      date: 12,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
  },
}
`;

snapshot[`parse() > taskがないときは終了時刻まで明記する 2`] = `
{
  ok: true,
  value: {
    end: {
      date: 12,
      hours: 14,
      minutes: 20,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな@2023-09-12T12:50/14:20",
    start: {
      date: 12,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
  },
}
`;

snapshot[`parse() > taskがないときは終了時刻まで明記する 3`] = `
{
  ok: false,
  value: {
    message: "No end date found in both a task specifier and an event specifier.",
    name: "LackDateError",
  },
}
`;

snapshot[`parse() > taskがないときは終了時刻まで明記する 4`] = `
{
  ok: false,
  value: {
    message: 'Could not found any "hours" parameter.',
    name: "LackDateError",
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 1`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      duration: 10,
      name: "ハイウェイ惑星 買おうかな@13",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13",
      start: {
        date: 12,
        hours: 12,
        minutes: 50,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 2`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12D10@13T12:50",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      duration: 10,
      name: "ハイウェイ惑星 買おうかな@13T12:50",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12D10@13T12:50",
      start: {
        date: 12,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 3`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50D10",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      name: "ハイウェイ惑星 買おうかな@13T12:50D10",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50D10",
      start: {
        date: 12,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 4`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 14,
      minutes: 20,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50/14:20",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      name: "ハイウェイ惑星 買おうかな@13T12:50/14:20",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50/14:20",
      start: {
        date: 12,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 5`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 12,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12@12-13T12:50D10",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 12,
      year: 2023,
    },
    task: {
      name: "ハイウェイ惑星 買おうかな@12-13T12:50D10",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12@12-13T12:50D10",
      start: {
        date: 12,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 6`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 12,
      year: 2024,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12@2024-12-13T12:50D10",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 12,
      year: 2024,
    },
    task: {
      name: "ハイウェイ惑星 買おうかな@2024-12-13T12:50D10",
      raw: "ハイウェイ惑星 買おうかな~@2023-09-12@2024-12-13T12:50D10",
      start: {
        date: 12,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 7`] = `
{
  ok: false,
  value: {
    message: 'Could not found any "hours" parameter.',
    name: "LackDateError",
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 8`] = `
{
  ok: false,
  value: {
    message: "No end date found in both a task specifier and an event specifier.",
    name: "LackDateError",
  },
}
`;

snapshot[`parse() > taskの日付情報から継承する 9`] = `
{
  ok: false,
  value: {
    message: 'Could not found any "hours" parameter.',
    name: "LackDateError",
  },
}
`;

snapshot[`parse() > taskとeventの日付指定は順不同 1`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "ハイウェイ惑星 買おうかな",
    raw: "@13ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10",
    start: {
      date: 13,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      duration: 10,
      name: "@13ハイウェイ惑星 買おうかな",
      raw: "@13ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10",
      start: {
        date: 12,
        hours: 12,
        minutes: 50,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;

snapshot[`parse() > taskとeventの日付指定は順不同 2`] = `
{
  ok: true,
  value: {
    end: {
      date: 24,
      hours: 13,
      minutes: 0,
      month: 9,
      year: 2023,
    },
    name: "←がTaskを指定し、~←がEventを指定する",
    raw: "~@2023-09-12T12:50D10 ←がTaskを指定し、~@09-24←がEventを指定する",
    start: {
      date: 24,
      hours: 12,
      minutes: 50,
      month: 9,
      year: 2023,
    },
    task: {
      duration: 10,
      name: "←がTaskを指定し、~@09-24←がEventを指定する",
      raw: "~@2023-09-12T12:50D10 ←がTaskを指定し、~@09-24←がEventを指定する",
      start: {
        date: 12,
        hours: 12,
        minutes: 50,
        month: 9,
        year: 2023,
      },
      status: "up-down",
    },
  },
}
`;
